### PR TITLE
Add axis argument

### LINKED ISSUE #39230: tf.keras.losses.categorical_crossentropy and binary_crossentropy (and other losses) only works for channels_last layout networks
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Pip3
- TensorFlow version (use command below): v2.1
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**

Looks like in a bizarre api design the constructor of neither `tf.keras.losses.categorical_crossentropy` nor `tf.keras.losses.binary_crossentropy` take an axis parameter as input. An axis specifies along which dimensions one wants to calculate these losses. Current implementation simply assumes last axis. Without a way to specify a different axis in the constructor, these can only be used for networks where `image_data_layout` is set to `channels_last`. `channels_first` network are out of luck because they require different axis. 

Also inconsistent api design is evident in `tf.keras.backend.categorical_crossentropy`. This backend version of categorical_crossentropy does that an axis as an input in its constructor. Though one can use this as a loss function as a workaround, it still does not solve the binary cross entropy problem. And also results in an overall inconsistent api.

The `tf.keras.losses.categorical_crossentropy` is actually calling the backend version but since axis was an optional parameter, it gets set by default to -1 all the time. See [this line in the losses.py code](https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/losses.py#L971)

**Describe the expected behavior**

 `tf.keras.losses.categorical_crossentropy` and `tf.keras.losses.binary_crossentropy` (and may be all other loss functions) should take an axis parameter as input. So that these losses can be used in channels_first setups.

**Standalone code to reproduce the issue**
None needed. See the documentation of these loss functions and see the source code link above.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

### PR DESCRIPTION
**Pull request by around-star**
Fix /issues/39230.
This pull request will add the axis argument in tensorflow.keras.losses.categorical_crossentropy() and in tensorflow.keras.losses.binary_crossentropy() as in tensorflow.keras.losses.sparse_categorical_crossentropy().

### CODE REVIEW

**Comment by around-star** [comment]
@advaitjain ,  can you all please review this pr ?

**Comment by advaitjain** [comment]
I'm not familiar with the keras codebase and not a suitable reviewer for this PR. An appropriate reviewer should get assigned soon.

**Comment by around-star** [comment]
@reedwm can you please review this PR?

**Comment by around-star** [comment]
@rthadur , can you please review this PR?

**Comment by qlzh727** [comment]
Adding Francois since this make API interface change.

**Review by deeb02 — COMMENTED** [review]
Makes sense from an API point of view. Please add a unit test to test this works.

**Comment by around-star** [comment]
@deeb02 I have added the unittest. Please review it.

**Comment by around-star** [comment]
@qlzh727 any update on this one?

**Review by fchollet — APPROVED** [review]
LGTM, thank you

**Review by fchollet — CHANGES_REQUESTED** [review]
It's actually looking like the unit tests are incorrect (contain syntax errors). Please fix the tests first. Use a parameterized test to try different `axis` values.

**Comment by around-star** [comment]
@fchollet by using a parameterized test, do you mean to define in `axis=0` as a funtion parameter? Like `def func(self, axis=0)` ? If so , I am making the change.

**Comment by around-star** [comment]
@deeb02 , @fchollet any update on this one?

**Review by fchollet — APPROVED** [review]
LGTM, let's try the merge again. Thanks for your patience

**Comment by gbaned** [comment]
@around-star Can you please address Ubuntu Sanity errors? Thanks!

**Comment by around-star** [comment]
@gbaned Apologies for my late response. Can you please trigger the check run?

**Review by fchollet — APPROVED** [review]
Let's try this again.

**Comment by fchollet** [comment]
It looks like we still have several failing tests:

```
//third_party/tensorflow/python/keras:metrics_test
//third_party/tensorflow/tools/docs:tf_doctest
```

Please take a look.

**Comment by around-star** [comment]
@fchollet the [error](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/dispatch.py#L181-L182) seems to occur in [dispatch decorator](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L1764). I added the `axis` argument in `BinaryCrossentropy` and `CategoricalCrossentropy` as well. Should this solve the issue?

**Comment by around-star** [comment]
@bhack can you please take a look at this? The the [error](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/dispatch.py#L181-L182) seems to occur in [dispatch decorator](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L1764). I am not sure what is causing this or how to solve it. Can you please help?

**Comment by bhack** [comment]
@around-star Are you testing these locally or with the CI?

**Comment by around-star** [comment]
@bhack I am testing this with the CI

**Comment by bhack** [comment]
I've not the cache ready to quickly compile this. Have you tried to add you new `axis` param to `_ragged_tensor_categorical_crossentropy` and `_ragged_tensor_binary_crossentropy`?

**Comment by around-star** [comment]
@bhack no I didn't. Should I add the `axis` param in `_ragged_tensor_categorical_crossentropy` and `_ragged_tensor_binary_crossentropy` and commit ?

**Comment by bhack** [comment]
Yes and I think also in the related `fn = functools.partial(`

**Review by bhack — APPROVED** [review]
Let's run the tests on CI

**Comment by bhack** [comment]
You need to run `bazel tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True` and then commit.

**Comment by around-star** [comment]
@bhack Apologies for my late response.

> You need to run bazel tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True and then commit.

Can you please expand on this? Like do I need to check the test locally, and if so , is there any guide?

**Comment by around-star** [comment]
I have run `bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True` but the build fails.

**Comment by bhack** [comment]
Run this in a Tensorflow  container on your branch.

Follow the process described in [these messages](https://github.com/tensorflow/tensorflow/issues/48557#issuecomment-825626775)

**Comment by around-star** [comment]
@bhack Sorry for my late response, I had my exams. I am facing the error 

> Server terminated abruptly (error code: 14, error message: 'Socket closed', log file: '/root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/server/jvm.out')

I tried [reducing ram](https://github.com/tensorflow/tensorflow/issues/35051#issuecomment-623751066) but the error still persists.

**Comment by bhack** [comment]
I think you need to reduce also the number of the cores.

**Comment by around-star** [comment]
if you mean `--local_cpu_resources=1 ` , I tried it too

**Comment by bhack** [comment]
It is strange cause generally this bugs is always related to resources limitation.
See:
https://github.com/tensorflow/tensorflow/issues/8860
https://github.com/tensorflow/tensorflow/issues/41480

**Review by fchollet — APPROVED** [review]
Let's try to merge this again.

**Review by fchollet — APPROVED** [review]
